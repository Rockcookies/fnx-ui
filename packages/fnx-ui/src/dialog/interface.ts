import { HTMLAttributes, ReactNode } from 'react';
import { ButtonProps } from '../button/interface';
import { PopupComponentProps } from '../popup/interface';

export interface DialogLocale {
	confirm: string;
	cancel: string;
}

export type DialogMessageAlign = 'left' | 'right' | 'center';

export interface DialogComponentProps
	extends Omit<PopupComponentProps, 'onClose'> {
	title?: ReactNode;
	footer?: ReactNode;
	width?: number | string;
	message?: ReactNode;
	messageAlign?: DialogMessageAlign;
	confirmButton?: ReactNode;
	confirmText?: string;
	confirmLoading?: boolean;
	confirmButtonProps?: ButtonProps;
	cancelButton?: ReactNode;
	cancelText?: string;
	cancelLoading?: boolean;
	cancelButtonProps?: ButtonProps;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
	onConfirm?: () => void;
	onCancel?: () => void;
}

export type DialogProps = DialogComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface DialogStaticProps extends DialogProps {
	onConfirm?: () => Promise<any> | any;
	onCancel?: () => Promise<any> | any;
}

export interface DialogInstance {
	update: (props: DialogProps) => void;
	clear: () => void;
}

export interface DialogContext {
	show: (props: DialogProps) => DialogInstance;
}

export interface DialogExportExtra {
	show: (props: DialogProps) => DialogInstance;
	useDialog: () => DialogContext;
	clearAll: () => void;
}
