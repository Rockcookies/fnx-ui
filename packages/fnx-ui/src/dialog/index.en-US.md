# Dialog

The map is pop-up, which is often used for message prompts, messages confirm, or complete specific interactions within the current page.

The pop-up frame component support function calls and components call two ways.

## Function Call

`Dialog.show` as a function, the corresponding modal frame is popped up directly to the page.

```tsx
import { Dialog } from 'fnx-ui';

Dialog.show({
  title: 'Dialog',
  message: 'Dialog Content',
  messageAlign: 'Left',
  confirmText: 'Confirm',
  cancelText: 'Cancel',
})ï¼›

```

## Component Call

When you call the `Dialog by the component, you can display hidden via` Visible` control.

```tsx
import { Dialog } from 'fnx-ui';

ReactDOM.render(
  <>
    <Dialog visible title="Dialog" message="Message" />
  </>,
  mountNode,
);
```

## Asynchronous close

Through `onConfirm`,`onCancel` property can be passed into a callback function, you need to perform it to turn it off, and return to `false`.

```tsx
import { Dialog } from 'fnx-ui';

function sleep(timeoutsMs) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, timeoutsMs);
  });
}

Dialog.show({
  title: 'Dialog',
  message: 'Dialog Content',
  messageAlign: 'Left',
  confirmText: 'Confirm',
  cancelText: 'Cancel',
  onConfirm: async () => {
    await sleep(1000);
  },
  onCancel: async () => {
    await sleep(1000);
    Toast.show('Stopped!');

    return false;
  },
});
```

## Customize

`Children` type is `ReactNode`, accept custom content.

```tsx
import { Dialog } from 'fnx-ui';

Dialog.show({
  title: 'Dialog',
  message: 'Dialog Content',
  messageAlign: 'Left',
  confirmText: 'Confirm',
  cancelText: 'Cancel',
  children: (
    <img src="https://img17.fn-mart.com/pic/2a021346939822ebb66f/B26n221z_2fdhMZdKz/7imyeafa3yKGj9/CsmRtmA3RAuAVmgsAAKtgt1PF3s715.jpg" />
  ),
});
```

## Dynamically Loading

The corresponding `Dialog` instance will be returned when performing a `Dialog` method, modify the `message` attribute on the instance via the` update` method to realize the effect of dynamic update prompts.

```tsx
import { Dialog, Button } from 'fnx-ui';
import React, { useRef } from 'react';

const dialog = Dialog.useDialog(); //DialogContext

const timerRef = useRef<NodeJS.Timeout>();

const handleCountdown = () => {
  let second = 3;
  const toast = dialog.show({
    message: `Now it is ${second}`,
  });

  timerRef.current = setInterval(() => {
    second--;

    if (second) {
      toast.update({
        message: `Now it is ${second}`,
      });
    } else {
      toast.clear();
    }
  }, 1000);
};

ReactDOM.render(
  <>
    <Button onClick={handleCountdown}>Dynamic Loading</Button>
  </>,
  mountNode,
);
```

## API

`Buttonprops` Type Reference` Button` components.

| Parameter           | Description                                     | Type                                 | Default value |
| ------------------- | ----------------------------------------------- | ------------------------------------ | ------------- |
| title               | title                                           | `ReactNode`                          | -             |
| footer              | Footage                                         | `ReactNode`                          | -             |
| width               | width                                           | `number \| string`                   | -             |
| message             | content                                         | `ReactNode`                          | -             |
| messageAlign        | Content alignment                               | `'left' \| 'right'`                  | `center`      |
| confirmButton       | Confirm button                                  | `ReactNode`                          | -             |
| confirmText         | Confirm button                                  | `string`                             | -             |
| confirmLoading      | Click to confirm if the load icon is displayed. | `boolean`                            | -             |
| confirmButtonProps  | Confirm button properties                       | `ButtonProps`                        | `{}`          |
| cancelButton        | Cancel button                                   | `ReactNode`                          | -             |
| cancelText          | Cancel button                                   | `string`                             | -             |
| cancelLoading       | Confirm that the load is completed              | `boolean`                            | -             |
| cancelButtonProps   | Cancel button properties                        | `ButtonProps`                        | `{}`          |
| bodyProps           | Content package layer properties                | `HTMLAttributes`                     | -             |
| visible             | Whether it is displayed                         | `boolean`                            | `false`       |
| round               | Whether to show the fillet                      | `boolean`                            | `false`       |
| position            | Location                                        | `` top` \| `bottom` \| `middle ``    | `middle`      |
| overlay             | Whether to display a mask layer                 | `boolean`                            | `fill`        |
| overlayClassName    | Mask layer class name                           | `string`                             | -             |
| overlayStyle        | Mask layer style                                | `string`                             | -             |
| overlayCloseable    | Whether to click the mask layer close           | `boolean`                            | `true`        |
| renderOnShow        | Whether to render load                          | `boolean`                            | `true`        |
| destroyOnHide       | Whether to close the destruction                | `boolean`                            | `false`       |
| mountTo             | Specify a mounted node                          | `boolean \| element \| ()=>element ` | -             |
| transitionDuration  | Transition animation                            | `number`                             | -             |
| transitionName      | Transition animation name                       | `string`                             | -             |
| lockScroll          | Whether it is forbidden to scroll               | `boolean`                            | `true`        |
| safeAreaInsetBottom | Do you open the bottom security area adaptation | `boolean`                            | `false`       |

## <Dialog /> Events

| Method Name    | Description                                       | Parameter                                        | Return Value |
| -------------- | ------------------------------------------------- | ------------------------------------------------ | ------------ |
| onConfirm      | Confirm callback                                  | -                                                | -            |
| onCancel       | Cancel                                            | -                                                | -            |
| onShow         | Trigger when opening the pop-up layer             | -                                                | -            |
| onOverlayClick | Click on the mask layer trigger                   | `(e: React.MouseEvent<HTMLElement, MouseEvent>)` | -            |
| onClose        | Mask layer can click on the mask layer to trigger | -                                                | -            |
| onBeforeHide   | Executive trigger before the animation            | -                                                | -            |
| onHide         | Executive trigger in the animation                | -                                                | -            |
| onAfterHide    | Executive trigger after the animation             | -                                                | -            |
| onBeforeShow   | Execute trigger before entering the field         | -                                                | -            |
| onShow         | Implementing triggers in the input animation      | -                                                | -            |
| onAfterShow    | Imported animation                                | -                                                | -            |

## Dialog Method

`Dialogprops` Type as component API.

| Method Name       | Description            | Parameter     | Return Value    |
| ----------------- | ---------------------- | ------------- | --------------- |
| Dialog.show()     | Display pop-up window  | `DialogProps` | Dialog Instance |
| Dialog.clearAll() | Destroy pop-up example | -             | -               |

### Dialog Instance Method

```
const useDialog = Dialog.useDialog();
const dialog=useDialog.show()
dialog.update();
dialog.clear();

```

| Method Name     | Description           | Parameter                                        | Return Value   |
| --------------- | --------------------- | ------------------------------------------------ | -------------- |
| dialog.update() | Update display prompt | `({ visible: boolean \| undefined, ...props }) ` | toast instance |
| dialog.clear()  | Destruction example   | -                                                | toast instance |
