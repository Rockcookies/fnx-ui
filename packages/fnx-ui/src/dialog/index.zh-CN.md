# Dialog 对话框

弹出模态框，常用于消息提示、消息确认，或在当前页面内完成特定的交互操作。

弹出框组件支持函数调用和组件调用两种方式。

## 函数调用

`Dialog.show` 作为函数，调用后会直接在页面中弹出相应的模态框。

```tsx
import { Dialog } from 'fnx-ui';

Dialog.show({
  title: 'Dialog',
  message: 'Dialog Content',
  messageAlign: 'Left',
  confirmText: 'Confirm',
  cancelText: 'Cancel',
})；

```

## 组件调用

通过组件调用 `Dialog` 时，可以通过 `visible` 控制显示隐藏。

```tsx
import { Dialog } from 'fnx-ui';

ReactDOM.render(
  <>
    <Dialog visible title="Dialog" message="Message" />
  </>,
  mountNode,
);
```

## 异步关闭

通过 `onConfirm` ，`onCancel` 属性可以传入一个回调函数，需要执行完才会关闭，不关闭则返回 `false`。

```tsx
import { Dialog } from 'fnx-ui';

function sleep(timeoutsMs) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, timeoutsMs);
  });
}

Dialog.show({
  title: 'Dialog',
  message: 'Dialog Content',
  messageAlign: 'Left',
  confirmText: 'Confirm',
  cancelText: 'Cancel',
  onConfirm: async () => {
    await sleep(1000);
  },
  onCancel: async () => {
    await sleep(1000);
    Toast.show('Stopped!');

    return false;
  },
});
```

## 自定义内容

`children` 类型是 `ReactNode`，接受自定义内容。

```tsx
import { Dialog } from 'fnx-ui';

Dialog.show({
  title: 'Dialog',
  message: 'Dialog Content',
  messageAlign: 'Left',
  confirmText: 'Confirm',
  cancelText: 'Cancel',
  children: (
    <img src="https://img17.fn-mart.com/pic/2a021346939822ebb66f/B26n221z_2fdhMZdKz/7imyeafa3yKGj9/CsmRtmA3RAuAVmgsAAKtgt1PF3s715.jpg" />
  ),
});
```

## 动态加载

执行 `Dialog` 方法时会返回对应的 `Dialog` 实例，通过 `update` 方法修改实例上的 `message` 属性可以实现动态更新提示的效果。

```tsx
import { Dialog, Button } from 'fnx-ui';
import React, { useRef } from 'react';

const dialog = Dialog.useDialog();

const timerRef = useRef<NodeJS.Timeout>();

const handleCountdown = () => {
  let second = 3;
  const toast = dialog.show({
    message: `Now it is ${second}`,
  });

  timerRef.current = setInterval(() => {
    second--;

    if (second) {
      toast.update({
        message: `Now it is ${second}`,
      });
    } else {
      toast.clear();
    }
  }, 1000);
};

ReactDOM.render(
  <>
    <Button onClick={handleCountdown}>Dynamic Loading</Button>
  </>,
  mountNode,
);
```

## API

`ButtonProps` 类型参考 `Button` 组件。

| 参数                | 说明                     | 类型                                 | 默认值   |
| ------------------- | ------------------------ | ------------------------------------ | -------- |
| title               | 标题                     | `ReactNode`                          | -        |
| footer              | 页脚                     | `ReactNode`                          | -        |
| width               | 宽度                     | `number \| string`                   | -        |
| message             | 内容                     | `ReactNode`                          | -        |
| messageAlign        | 内容对齐方式             | `'left' \| 'right'`                  | `center` |
| confirmButton       | 确认按钮                 | `ReactNode`                          | -        |
| confirmText         | 确认按钮文案             | `string`                             | -        |
| confirmLoading      | 点击确认是否显示加载图标 | `boolean`                            | -        |
| confirmButtonProps  | 确认按钮属性             | `ButtonProps`                        | `{}`     |
| cancelButton        | 取消按钮                 | `ReactNode`                          | -        |
| cancelText          | 取消按钮文案             | `string`                             | -        |
| cancelLoading       | 确认加载完毕             | `boolean`                            | -        |
| cancelButtonProps   | 取消按钮属性             | `ButtonProps`                        | `{}`     |
| bodyProps           | 内容包裹层属性           | `HTMLAttributes`                     | -        |
| visible             | 是否显示弹出层           | `boolean`                            | `false`  |
| round               | 是否展示圆角             | `boolean`                            | `false`  |
| position            | 位置                     | `` top` \| `bottom` \| `middle ``    | `middle` |
| overlay             | 是否显示遮罩层           | `boolean`                            | `fill`   |
| overlayClassName    | 遮罩层类名               | `string`                             | -        |
| overlayStyle        | 遮罩层样式               | `string`                             | -        |
| overlayCloseable    | 是否点击遮罩层关闭       | `boolean`                            | `true`   |
| renderOnShow        | 是否渲染加载             | `boolean`                            | `true`   |
| destroyOnHide       | 是否关闭销毁             | `boolean`                            | `false`  |
| mountTo             | 指定挂载的节点           | `boolean \| element \| ()=>element ` | -        |
| transitionDuration  | 过渡动画                 | `number`                             | -        |
| transitionName      | 过渡动画类名             | `string`                             | -        |
| lockScroll          | 是否禁止滚动             | `boolean`                            | `true`   |
| safeAreaInsetBottom | 是否开启底部安全区适配   | `boolean`                            | `false`  |

## <Dialog /> 事件

| 方法名         | 说明                         | 参数                                             | 返回值 |
| -------------- | ---------------------------- | ------------------------------------------------ | ------ |
| onConfirm      | 确认回调                     | -                                                | -      |
| onCancel       | 取消回调                     | -                                                | -      |
| onShow         | 打开弹出层时触发             | -                                                | -      |
| onOverlayClick | 点击遮罩层触发               | `(e: React.MouseEvent<HTMLElement, MouseEvent>)` | -      |
| onClose        | 遮罩层可点击时点击遮罩层触发 | -                                                | -      |
| onBeforeHide   | 出场动画前执行触发           | -                                                | -      |
| onHide         | 出场动画中执行触发           | -                                                | -      |
| onAfterHide    | 出场动画后执行触发           | -                                                | -      |
| onBeforeShow   | 进场动画前执行触发           | -                                                | -      |
| onShow         | 进场动画中执行触发           | -                                                | -      |
| onAfterShow    | 进场动画后执行触发           | -                                                | -      |

## Dialog 方法

`DialogProps`类型为组件 API。

| 方法名            | 说明         | 参数          | 返回值   |
| ----------------- | ------------ | ------------- | -------- |
| Dialog.show()     | 展示弹窗     | `DialogProps` | 弹窗示例 |
| Dialog.clearAll() | 销毁弹窗实例 | -             | -        |

### Dialog 实例方法

```
const useDialog = Dialog.useDialog();
const dialog=useDialog.show()
dialog.update();
dialog.clear();

```

| 方法名          | 说明         | 参数                                             | 返回值     |
| --------------- | ------------ | ------------------------------------------------ | ---------- |
| dialog.update() | 更新展示提示 | `({ visible: boolean \| undefined, ...props }) ` | toast 实例 |
| dialog.clear()  | 销毁实例     | -                                                | toast 实例 |
