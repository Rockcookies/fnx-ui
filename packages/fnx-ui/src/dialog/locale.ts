import defaultLocale from '../locale/default';
import { DialogLocale } from './interface';

// thanks to: https://github.com/ant-design/ant-design/blob/HEAD/components/modal/locale.tsx

let runtimeLocale: DialogLocale = {
	...defaultLocale.dialog,
};

export function changeConfirmLocale(newLocale?: DialogLocale) {
	if (newLocale) {
		runtimeLocale = {
			...runtimeLocale,
			...newLocale,
		};
	} else {
		runtimeLocale = {
			...defaultLocale.dialog,
		};
	}
}

export function getConfirmLocale(): DialogLocale {
	return runtimeLocale;
}
