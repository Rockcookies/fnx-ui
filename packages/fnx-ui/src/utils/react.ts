import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
