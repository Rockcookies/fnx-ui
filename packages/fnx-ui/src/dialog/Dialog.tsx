import React, { forwardRef } from 'react';
import Button, { ButtonProps } from '../button';
import configComponentProps from '../hooks/config-component-props';
import { useLocale } from '../locale';
import Popup from '../popup';
import { BORDER_LEFT, BORDER_TOP } from '../utils/constants';
import { addUnit } from '../utils/format';
import { noop } from '../utils/misc';
import { classnames, createBEM } from '../utils/namespace';
import { DialogProps } from './interface';

const NS = 'fnx-dialog';
const bem = createBEM(NS);

const useProps = configComponentProps<
	Required<
		Pick<
			DialogProps,
			| 'messageAlign'
			| 'onConfirm'
			| 'onCancel'
			| 'confirmLoading'
			| 'confirmButtonProps'
			| 'cancelLoading'
			| 'cancelButtonProps'
			| 'transitionName'
			| 'bodyProps'
		>
	>
>({
	messageAlign: 'center',
	onConfirm: noop,
	onCancel: noop,
	confirmLoading: false,
	confirmButtonProps: {},
	cancelLoading: false,
	cancelButtonProps: {},
	transitionName: 'fnx-dialog-bounce',
	bodyProps: {},
});

const Dialog = forwardRef<HTMLDivElement, DialogProps>((_props, ref) => {
	const locale = useLocale('dialog');

	const [
		{
			messageAlign,
			onConfirm,
			onCancel,
			confirmLoading,
			cancelLoading,
			confirmButtonProps,
			cancelButtonProps,
			transitionName,
			bodyProps,
		},
		{
			title,
			footer,
			width,
			message,
			confirmButton,
			confirmText,
			cancelButton,
			cancelText,
			style,
			className,
			children,
			...resetProps
		},
	] = useProps(_props);

	const hasHeader = !!title;
	const hasBody = !!message || !!children;
	const hasFooter = footer !== false;

	const renderHeader = () => {
		return (
			<div
				className={bem('header', {
					isolated: !hasBody,
				})}
			>
				{title}
			</div>
		);
	};

	const renderBody = () => {
		if (children) {
			return (
				<div
					{...bodyProps}
					className={classnames(bem('body'), bodyProps.className)}
				>
					{children}
				</div>
			);
		}

		if (message) {
			return (
				<div
					className={bem('message', {
						'has-header': hasHeader,
						[messageAlign]: messageAlign,
					})}
				>
					{message}
				</div>
			);
		}
	};

	const renderButton = ({
		loading,
		children,
		onClick,
		className,
		extraProps,
	}: ButtonProps & { extraProps: ButtonProps }) => {
		return (
			<Button
				loading={loading}
				{...extraProps}
				className={classnames(className, extraProps.className)}
				onClick={(...args) => {
					onClick && onClick(...args);
					extraProps.onClick && extraProps.onClick(...args);
				}}
			>
				{children}
			</Button>
		);
	};

	const renderFooter = () => {
		if (footer) {
			return (
				<div className={classnames(BORDER_TOP, bem('footer'))}>
					{footer}
				</div>
			);
		}

		return (
			<div className={classnames(BORDER_TOP, bem('footer'))}>
				{confirmButton !== false &&
					(confirmButton ||
						renderButton({
							loading: confirmLoading,
							children: confirmText || locale.confirm,
							className: bem('confirm'),
							extraProps: confirmButtonProps,
							onClick: onConfirm,
						}))}
				{cancelButton !== false &&
					(cancelButton ||
						renderButton({
							loading: cancelLoading,
							children: cancelText || locale.cancel,
							className: classnames(
								confirmButton !== false
									? BORDER_LEFT
									: undefined,
								bem('cancel'),
							),
							extraProps: cancelButtonProps,
							onClick: onCancel,
						}))}
			</div>
		);
	};

	return (
		<Popup
			role="dialog"
			className={classnames(bem(), className)}
			style={{ width: addUnit(width), ...style }}
			ref={ref}
			overlayCloseable={false}
			transitionName={transitionName}
			{...resetProps}
			onClose={onCancel}
		>
			{hasHeader && renderHeader()}
			{hasBody && renderBody()}
			{hasFooter && renderFooter()}
		</Popup>
	);
});

Dialog.displayName = 'Dialog';

export default Dialog;
