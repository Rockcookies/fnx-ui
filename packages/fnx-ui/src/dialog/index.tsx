import React, { FC, useContext, useEffect, useMemo, useState } from 'react';
import { ConfigContext } from '../config-provider/context';
import useDefaultsRef from '../hooks/use-defaults-ref';
import PopupHelper from '../popup/utils/popup-helper';
import { isFunction, isPromise } from '../utils/detect';
import { noop } from '../utils/misc';
import CDialog from './Dialog';
import {
	DialogContext,
	DialogExportExtra,
	DialogInstance,
	DialogLocale,
	DialogStaticProps,
} from './interface';
import { getConfirmLocale } from './locale';

const helper = new PopupHelper();

const createDialog = (
	baseProps: DialogStaticProps,
	options: { locale?: DialogLocale; container?: HTMLElement } = {},
): DialogInstance => {
	return helper.create<DialogStaticProps>(({ onUpdate, remove }) => {
		const DialogStatic: FC<DialogStaticProps> = () => {
			const [{ onConfirm, onCancel, onAfterHide, ...props }, setProps] =
				useState(baseProps);

			const locale = options.locale || getConfirmLocale();

			const handleAction = (
				loadingKey: keyof DialogStaticProps,
				action: any,
			) => {
				if (!isFunction(action)) {
					action = noop;
				}

				const res = action();

				if (isPromise(res)) {
					setProps((prev) => ({ ...prev, [loadingKey]: true }));
					res.then((flag) => {
						setProps((prev) => ({
							...prev,
							[loadingKey]: false,
							visible: flag === false,
						}));
					}).catch(() => {
						setProps((prev) => ({
							...prev,
							[loadingKey]: false,
						}));
					});
				} else {
					setProps((prev) => ({ ...prev, visible: res === false }));
				}
			};

			useEffect(() => {
				const listener = (next: DialogStaticProps) => {
					setProps((prev) => ({ ...prev, ...next }));
				};

				return onUpdate(listener);
			}, []);

			return (
				<CDialog
					visible={true}
					{...props}
					mountTo={false}
					confirmText={props.confirmText || locale.confirm}
					cancelText={props.cancelText || locale.cancel}
					onConfirm={() => handleAction('confirmLoading', onConfirm)}
					onCancel={() => handleAction('cancelLoading', onCancel)}
					onAfterHide={() => {
						remove();
						onAfterHide && onAfterHide();
					}}
				/>
			);
		};

		DialogStatic.displayName = 'Dialog';

		return DialogStatic;
	}, options.container);
};

const Dialog: typeof CDialog & DialogExportExtra = CDialog as any;

Dialog.show = (props: DialogStaticProps) => {
	return createDialog(props);
};

Dialog.clearAll = () => {
	helper.clear();
};

Dialog.useDialog = () => {
	const context = useContext(ConfigContext);
	const contextRef = useDefaultsRef(context);

	return useMemo<DialogContext>(() => {
		return {
			show: (props) => {
				const { locale, mountTo } = contextRef.current;
				return createDialog(props, {
					locale: locale.dialog,
					container: mountTo(),
				});
			},
		};
	}, [contextRef]);
};

export type {
	DialogComponentProps,
	DialogProps,
	DialogContext,
	DialogInstance,
	DialogLocale,
	DialogMessageAlign,
	DialogStaticProps,
} from './interface';

export default Dialog;
