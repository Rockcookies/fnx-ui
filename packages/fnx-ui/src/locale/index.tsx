import { useContext, useMemo } from 'react';
import ConfigProvider from '../config-provider';
import { Dictionary } from '../utils/interface';

export function useLocale(key: string): Dictionary {
	const { locale } = useContext(ConfigProvider.Context);

	return useMemo(() => {
		const output: any = locale || {};
		return output[key] || {};
	}, [locale, key]);
}
