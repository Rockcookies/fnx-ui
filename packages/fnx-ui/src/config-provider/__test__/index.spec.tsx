import { render } from '@testing-library/react';
import React, { useContext } from 'react';
import ConfigProvider from '../';
import { getConfirmLocale } from '../../dialog/locale';
import useCreation from '../../hooks/use-creation';
import enUS from '../../locale/en-US';
import zhCN from '../../locale/zh-CN';

describe('<ConfigProvider />', () => {
	it('test provider', async () => {
		function Test() {
			const context = useContext(ConfigProvider.Context);

			return (
				<pre>
					{JSON.stringify({
						...context,
						locale: context.locale.locale,
					})}
				</pre>
			);
		}

		expect(getConfirmLocale()).toEqual(enUS.dialog);

		const { container } = render(
			<ConfigProvider
				locale={zhCN}
				transitionDuration={600}
				safeAreaInsetBottom={false}
			>
				<Test />
			</ConfigProvider>,
		);

		expect(getConfirmLocale()).toEqual(zhCN.dialog);
		expect(container).toMatchSnapshot();
	});

	it('test-inheritance', async () => {
		function Test() {
			const context = useContext(ConfigProvider.Context);

			const count = useCreation<number>(
				(prev) => {
					return (prev || 0) + 1;
				},
				[...Object.values(context)],
			);

			return (
				<pre>
					{JSON.stringify({
						count,
						...context,
						locale: context.locale.locale,
					})}
				</pre>
			);
		}

		const { container, rerender } = render(
			<ConfigProvider
				locale={zhCN}
				transitionDuration={600}
				safeAreaInsetBottom={true}
			>
				<ConfigProvider locale={zhCN} transitionDuration={500}>
					<Test />
				</ConfigProvider>
			</ConfigProvider>,
		);

		expect(container).toMatchSnapshot();

		rerender(
			<ConfigProvider locale={zhCN} safeAreaInsetBottom={true}>
				<ConfigProvider locale={zhCN} transitionDuration={500}>
					<Test />
				</ConfigProvider>
			</ConfigProvider>,
		);

		expect(container).toMatchSnapshot();
	});
});
