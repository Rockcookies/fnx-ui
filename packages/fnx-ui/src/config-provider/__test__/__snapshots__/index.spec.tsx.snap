// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfigProvider /> test provider 1`] = `
<div>
  <pre>
    {"locale":"zh-CN","safeAreaInsetBottom":false,"transitionDuration":600}
  </pre>
</div>
`;

exports[`<ConfigProvider /> test-inheritance 1`] = `
<div>
  <pre>
    {"count":1,"locale":"zh-CN","safeAreaInsetBottom":true,"transitionDuration":500}
  </pre>
</div>
`;

exports[`<ConfigProvider /> test-inheritance 2`] = `
<div>
  <pre>
    {"count":1,"locale":"zh-CN","safeAreaInsetBottom":true,"transitionDuration":500}
  </pre>
</div>
`;
